cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1616,52 +1657,70 @@ void reader::impl::allocate_columns(size_t skip_rows, size_t num_rows, bool uses
 
   // compute output column sizes by examining the pages of the -input- columns
   if (has_lists) {
-    auto& page_keys  = _chunk_itm_data.page_keys;
     auto& page_index = _chunk_itm_data.page_index;
+
+    hostdevice_vector<input_col_info> input_cols{_input_columns.size(), _stream};
+    size_t max_depth = 0;
+    for (size_t i = 0; i < _input_columns.size(); i++) {
+      auto depth                  = _input_columns[i].nesting_depth();
+      max_depth                   = depth > max_depth ? depth : max_depth;
+      input_cols[i].nesting_depth = depth;
+      input_cols[i].schema_idx    = _input_columns[i].schema_idx;
+    }
+    input_cols.host_to_device(_stream);
+
+    // size iterator. indexes pages by sorted order
+    auto size_input = cudf::detail::make_counting_transform_iterator(
+      0,
+      get_page_nesting_size{
+        input_cols.device_ptr(), max_depth, pages.size(), pages.device_ptr(), page_index.begin()});
+
+    auto reduction_keys =
+      cudf::detail::make_counting_transform_iterator(0, get_reduction_key{pages.size()});
+    hostdevice_vector<size_t> sizes{_input_columns.size() * max_depth, _stream};
+    auto num_keys = _input_columns.size() * max_depth * pages.size();
+
+    // find the size of each column
+    thrust::reduce_by_key(rmm::exec_policy(_stream),
+                          reduction_keys,
+                          reduction_keys + num_keys,
+                          size_input,
+                          thrust::make_discard_iterator(),
+                          sizes.d_begin());
+
+    // for nested hierarchies, compute per-page start offset
+    thrust::exclusive_scan_by_key(rmm::exec_policy(_stream),
+                                  reduction_keys,
+                                  reduction_keys + num_keys,
+                                  size_input,
+                                  start_offset_output_iterator{pages.device_ptr(),
+                                                               page_index.begin(),
+                                                               0,
+                                                               input_cols.device_ptr(),
+                                                               max_depth,
+                                                               pages.size()});
```

Comment:
I wonder if we can even do better by combining these 2 kernel calls? Since they are operating on `reduction_keys`. Maybe just one `reduce_by_key` with a custom device lambda/functor that can do both reduce and scan?

Reply:
I'm not quite sure how to create a functor for `reduce_by_key` that would also perform the scan but it could be possible. Do you have an idea in mind?